File: backend/app/schemas/notification.py
"""Pydantic schemas for Notification API requests/responses."""

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, field_validator


class NotificationType(str, Enum):
    """Enum for notification types."""

    assignment_created = "assignment_created"
    deadline_approaching = "deadline_approaching"
    feedback_received = "feedback_received"
    message_received = "message_received"
    student_completed = "student_completed"
    past_due = "past_due"
    material_shared = "material_shared"
    system_announcement = "system_announcement"


class NotificationBase(BaseModel):
    """Base schema for notification data."""

    title: str = Field(..., min_length=1, max_length=500)
    message: str = Field(..., min_length=1)
    link: str | None = Field(None, max_length=500)

    @field_validator("title")
    @classmethod
    def validate_title_not_empty(cls, v: str) -> str:
        """Validate title is not empty or whitespace."""
        if not v.strip():
            raise ValueError("Title cannot be empty")
        return v


class NotificationCreate(NotificationBase):
    """Schema for creating a notification (service layer usage)."""

    user_id: uuid.UUID
    type: NotificationType


class NotificationResponse(BaseModel):
    """Schema for notification API response."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    type: NotificationType
    title: str
    message: str
    link: str | None
    is_read: bool
    created_at: datetime


class NotificationListResponse(BaseModel):
    """Schema for paginated notification list response."""

    notifications: list[NotificationResponse]
    total: int
    limit: int
    offset: int
    has_more: bool


class UnreadCountResponse(BaseModel):
    """Schema for unread notification count response."""

    count: int


class MarkAllReadResponse(BaseModel):
    """Schema for mark all as read response."""

    marked_count: int

File: frontend/src/types/notification.ts
/**
 * Notification type definitions for Dream LMS.
 * Story 6.1: Notification System Foundation
 * Story 6.5: Feedback Enhancements (Badge names in feedback notifications)
 */

/**
 * Notification type enumeration matching backend
 */
export type NotificationType =
  | "assignment_created"
  | "deadline_approaching"
  | "feedback_received"
  | "message_received"
  | "student_completed"
  | "past_due"
  | "material_shared"
  | "system_announcement"

/**
 * Notification interface with all fields
 */
export interface Notification {
  id: string
  user_id: string
  type: NotificationType
  title: string
  message: string
  link: string | null
  is_read: boolean
  created_at: string // ISO 8601 datetime string
}

/**
 * Paginated notification list response from API
 */
export interface NotificationListResponse {
  notifications: Notification[]
  total: number
  limit: number
  offset: number
  has_more: boolean
}

/**
 * Unread count response from API
 */
export interface UnreadCountResponse {
  count: number
}

/**
 * Mark all as read response from API
 */
export interface MarkAllReadResponse {
  marked_count: number
}

/**
 * Query parameters for fetching notifications
 */
export interface NotificationQueryParams {
  unread_only?: boolean
  type?: NotificationType
  limit?: number
  offset?: number
}

/**
 * Notification icon mapping type
 */
export type NotificationIconName =
  | "FileText" // assignment_created
  | "Clock" // deadline_approaching
  | "MessageSquare" // feedback_received
  | "Mail" // message_received
  | "CheckCircle" // student_completed
  | "AlertTriangle" // past_due
  | "Share2" // material_shared
  | "Bell" // system_announcement

// =============================================================================
// Notification Preferences (Story 6.8)
// =============================================================================

/**
 * Single notification preference
 */
export interface NotificationPreference {
  notification_type: NotificationType
  enabled: boolean
  email_enabled: boolean
  label: string
  description: string
}

/**
 * Global mute status
 */
export interface GlobalMuteStatus {
  muted_until: string // ISO 8601 datetime
  remaining_hours: number
}

/**
 * Response for preferences list with optional mute status
 */
export interface NotificationPreferencesResponse {
  preferences: NotificationPreference[]
  global_mute: GlobalMuteStatus | null
}

/**
 * Request for bulk updating preferences
 */
export interface NotificationPreferencesBulkUpdate {
  preferences: Record<NotificationType, boolean>
}

/**
 * Response for bulk preference update
 */
export interface NotificationPreferencesBulkUpdateResponse {
  updated: string[]
  preferences: NotificationPreference[]
}

/**
 * Request for single preference update
 */
export interface NotificationPreferenceUpdate {
  enabled: boolean
}

/**
 * Request to set global mute
 */
export interface GlobalMuteRequest {
  hours: number // 1-24
}

/**
 * Response for mute status check
 */
export interface GlobalMuteStatusResponse {
  is_muted: boolean
  mute: GlobalMuteStatus | null
}
